src/fides/ctl/core/dataset.py

```diff
@@ -37,6 +37,9 @@ def get_all_server_datasets(
         exclude_keys=exclude_dataset_keys,
         headers=headers,
     )
+    if not dataset_list:
+        return None
```

Comment:
With a signature of `List[Dataset]` I would be surprised to get an empty list. `Optional[List[Dataset]]` and returning `None` makes it clear the returned value needs to be checked to see if it has a value.

Reply:
I disagree here. A return type assertion of `List[Dataset]` implies nothing about the length of the returned list, so for a call to return an empty list is fine. Meanwhile, `Optional[List[Dataset]]` requires callers to take extra measures to handle the possibility of the function returning `None`, and that assumes that the engineer who implements the call reads the documented signature. In general, I strongly prefer that functions have a single return type whenever possible - it greatly reduces the likelihood that the function goes misused.

Reply:
I disagree in the other direction. If the return type is `List[Dataset]` I would expect it to have at least one dataset. Returning `[]` makes it ambiguous and error prone.

Does this work? The answer is maybe, maybe not. Based on the type I would expect it to, but I'm setup for errors

```py
def get_all_server_datasets() -> List[Dataset]:
    """Return either [] or [dataset]."""
    return ...

x = get_all_server_datasets()
print(x[0])
```

But this works every time without error

```py
def get_all_server_datasets() -> Optional[List[Dataset]]:
    """Return either None or [dataset]."""
    return ...

x = get_all_server_datasets()
if not x:
    print("I just saved you from an error")

print(x[0])
```

Could you do the same check by checking the length of the return instead of checking for `None`? Sure, but again it's ambiguous. Maybe you need to, maybe you don't. With optional you know without question you need to check.

Reply:
You are making assumptions you want to iterate the whole list, maybe you only want a specific index…see, assumptions are bad 😂

Honestly your stance surprises me with your love of Go. Rust wouldn’t even compile with you example, it would force you to return none instead of an empty vec. Would Go allow this?

Reply:
> You are making assumptions you want to iterate the whole list, maybe you only want a specific index…see, assumptions are bad 😂

In your example you wanted the first index only. I'll admit I overlooked that. Here:
```python
def get_all_server_datasets() -> List[Dataset]:
    """Always return a list"""
    return ...

datasets = get_all_server_datasets()
print(datasets[0]) if len(datasets) > 0 else print("I just saved you from an error")
# or
for dataset in datasets[0:1]:
    print(dataset)
```
To your credit, you did say that you'd have to check the length of the list, but I'd argue that defensive code would do that _even if_ `get_all_server_datasets` can potentially return `None`.

> Would Go allow this?

This Go code would compile:
```go
func test() []string {
    return nil
}
```
Because `nil` is the "zero value" for a slice. But, if you provided that, you'd be crucified. The idiomatic code would look like:
```go
func test() []string {
    return []string{}
}
```

Reply:
Thinking more about it, the _even better_ Go code would probably look like:
```go
func test() ([]string, error) {
    return nil, fmt.Errorf("Nothing to return")
}
```
Ugh. It's been way too long since I've written Go ❤️

Reply:
to play someone's advocate here...an empty list acts the same as `None` when used in a conditional, so if you were programming defensively than they look identical

```python
In [3]: foo = []

In [4]: if foo:
   ...:     print(foo[0])
   ...:
```
no issue here

Reply:
What in the world...

Reply:
empty collections are all treated that way 🤷 , the conditional is basically doing `if len(foo) > 0`

Reply:
further reading about Python dunder methods, this one in particular: https://blog.finxter.com/python-__len__-magic-method/

Reply:
No argument there, the `if foo` works for sure, but it is the same as checking the length in the end. It assumes someone will check it, where optional forces you to check it. “Forces” is a bit strong I guess because you could ignore the type checker, but you have been warned so if it blows up its your own fault. In the other examples it blows up and you don’t see it coming.